### PR TITLE
fix(tabs_bar): Allow animation to end before navigating

### PR DESCRIPTION
**Pull request by sorin-davidoi**
Another hacky solution, similar to #4320.

cc: @unarist @nolanlawson

### CODE REVIEW

**Comment by nolanlawson** [comment]
This solution looks great to me! Is there really not a "React-ish" solution though, like what Twitter describes [here](https://medium.com/@paularmstrong/twitter-lite-and-high-performance-react-progressive-web-apps-at-scale-d28a00e780a3) under "Defer Rendering When Mounting & Unmounting Many Components"? I kinda figured we could just use their same approach of deferring the component.

**Review by nolanlawson — APPROVED** [review]

File: app/javascript/mastodon/features/ui/components/tabs_bar.js [hunk 1]
```diff
@@ -25,16 +26,46 @@ export function getLink (index) {
...
+      const currentTab = this.node.querySelector('.tabs-bar__link.active');
+      const nextTab = e.nativeEvent.path.find(node => node.classList.contains('tabs-bar__link'));
+      const { props: { to } } = links[Array(...this.node.childNodes).indexOf(nextTab)];
+
+      currentTab.classList.remove('active');
+
+      const listener = debounce(() => {
+        nextTab.removeEventListener('transitionend', listener);
+        this.context.router.history.push(to);
+      }, 50);
```
Comment (nolanlawson) [inline app/javascript/mastodon/features/ui/components/tabs_bar.js]: What's with the debounce? `transitionend` should only be called once, exactly when the transition has ended.

File: app/javascript/mastodon/features/ui/components/tabs_bar.js [hunk 2]
```diff
@@ -25,16 +26,46 @@ export function getLink (index) {
...
+    this.node = ref;
+  }
+
+  handleClick = (e) => {
+    e.preventDefault();
+    e.persist();
+
+    requestAnimationFrame(() => {
+      const currentTab = this.node.querySelector('.tabs-bar__link.active');
+      const nextTab = e.nativeEvent.path.find(node => node.classList.contains('tabs-bar__link'));
```
Comment (unarist) [inline app/javascript/mastodon/features/ui/components/tabs_bar.js]: `path` property is not exist in Firefox and Safari.

**Review by unarist — CHANGES_REQUESTED** [review]

File: app/javascript/mastodon/features/ui/components/tabs_bar.js (hunk 1 above)
Comment (sorin-davidoi) [inline app/javascript/mastodon/features/ui/components/tabs_bar.js]: It has `transition: all`, and that class changes like 4 properties, all firing events.

File: app/javascript/mastodon/features/ui/components/tabs_bar.js (hunk 2 above)
Comment (sorin-davidoi) [inline app/javascript/mastodon/features/ui/components/tabs_bar.js]: Today I learned :+1:

**Review by unarist — APPROVED** [review]

**Comment by sorin-davidoi** [comment]
@nolan Will give it a shot but last time I checked the biggest bottleneck was not the rendering but the unmounting of the all the component on the page.

**Comment by unarist** [comment]
> Another problem is the animation. We would have to delay the mounting of the component not by animation frame, but by the duration of the transition (200ms).

I've missed this point, and yes, this change looks more laggy on the PC...

**Comment by sorin-davidoi** [comment]
We could wrap it inside isUserTouching, like the actions modal.

**Comment by unarist** [comment]
Although I don't think that feeling is not depending on whether it's PC or mobile device, but device performance or something.

**Comment by sorin-davidoi** [comment]
Yes, but I would say that whether the device is touch or not is strongly correlated to the performance of the device. Don't know what to say, let's leave it as it is for now and hear more feedback?

**Comment by unarist** [comment]
I found a odd behavior.

1. Open home column
2. Touch home tab, and just after, touch notification tab
3. It will show notification column in a moment, then back to home column

<details>
<summary>demo gif (w/ event timing log)</summary>

test
</details>

**Pull request merged**